Added CI action for publishing to maven repository

martin-gaievski opened the pull request:
### Description
Need CI action to publish artifacts to maven repo

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

heemin32 reviewed (approved):
LGTM. Thanks!

peterzhuamazon reviewed (commented):
Hi @martin-gaievski 
A few comments:
1. Remove line 27/28 and add the `all project` section from https://github.com/opensearch-project/user-behavior-insights/pull/98 to set `version` and group properly
2. opensearch-search-relevance need to be updated in settings.gradle
3. Add similar scripts/build.sh based on above PR and make sure all tasks called is actually in search relavance by using `./gradlew tasks`.

Thanks.

martin-gaievski commented:
> Hi @martin-gaievski A few comments:
> 
> 1. Remove line 27/28 and add the `all project` section from [Resolve incorrect maven and SNAPSHOT setup in UBI user-behavior-insights#98](https://github.com/opensearch-project/user-behavior-insights/pull/98) to set `version` and group properly
> 2. opensearch-search-relevance need to be updated in settings.gradle
> 3. Add similar scripts/build.sh based on above PR and make sure all tasks called is actually in search relavance by using `./gradlew tasks`.
> 
> Thanks.

Addressed comments, this is what I got after running `./gradle tasks`

```
./gradlew tasks
Starting a Gradle Daemon, 1 busy Daemon could not be reused, use --status for details
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 8.10.2
  OS Info               : Mac OS X 15.5 (aarch64)
  JDK Version           : 21 (Amazon Corretto JDK)
  JAVA_HOME             : /Users/gaievski/.sdkman/candidates/java/21.0.3-amzn
  Random Testing Seed   : 55B521E2AAB76F2A
  Crypto Standard       : any-supported
=======================================

> Task :tasks

------------------------------------------------------------
Tasks runnable from root project 'opensearch-search-relevance' - opensearch plugin that enables users to fine tuning search based on search relevance judgments
------------------------------------------------------------

Build tasks
-----------
assemble - Assembles the outputs of this project.
build - Assembles and tests this project.
buildDependents - Assembles and tests this project and all projects that depend on it.
buildNeeded - Assembles and tests this project and all projects it depends on.
classes - Assembles main classes.
clean - Deletes the build directory.
jar - Assembles a jar archive containing the classes of the 'main' feature.
javadocJar - Assembles a jar containing javadocs.
sourcesJar - Assembles a jar containing source files.
testClasses - Assembles test classes.
testFixturesClasses - Assembles test fixtures classes.
testFixturesJar - Assembles a jar archive containing the classes of the 'testFixtures' feature.
yamlRestTestClasses - Assembles yaml rest test classes.

Build Setup tasks
-----------------
init - Initializes a new Gradle build.
updateDaemonJvm - Generates or updates the Gradle Daemon JVM criteria.
wrapper - Generates Gradle wrapper files.

Documentation tasks
-------------------
javadoc - Generates Javadoc API documentation for the 'main' feature.
testFixturesJavadoc - Generates Javadoc API documentation for the 'testFixtures' feature.

Help tasks
----------
buildEnvironment - Displays all buildscript dependencies declared in root project 'opensearch-search-relevance'.
dependencies - Displays all dependencies declared in root project 'opensearch-search-relevance'.
dependencyInsight - Displays the insight into a specific dependency in root project 'opensearch-search-relevance'.
help - Displays a help message.
javaToolchains - Displays the detected java toolchains.
outgoingVariants - Displays the outgoing variants of root project 'opensearch-search-relevance'.
projects - Displays the sub-projects of root project 'opensearch-search-relevance'.
properties - Displays the properties of root project 'opensearch-search-relevance'.
resolvableConfigurations - Displays the configurations that can be resolved in root project 'opensearch-search-relevance'.
tasks - Displays the tasks runnable from root project 'opensearch-search-relevance'.

IDE tasks
---------
cleanEclipse - Cleans all Eclipse files.
cleanIdea - Cleans IDEA project files (IML, IPR)
eclipse - Generates all Eclipse files.
idea - Generates IDEA project files (IML, IPR, IWS)
openIdea - Opens the IDEA project

Lombok tasks
------------
delombok - Runs delombok on the main source-set
delombokTest - Runs delombok on the test source-set
delombokTestFixtures - Runs delombok on the testFixtures source-set
delombokYamlRestTest - Runs delombok on the yamlRestTest source-set
generateEffectiveLombokConfig - Generate effective Lombok configuration for source-set 'main'.
generateTestEffectiveLombokConfig - Generate effective Lombok configuration for source-set 'test'.
generateTestFixturesEffectiveLombokConfig - Generate effective Lombok configuration for source-set 'testFixtures'.
generateYamlRestTestEffectiveLombokConfig - Generate effective Lombok configuration for source-set 'yamlRestTest'.

OpenSearch cluster formation tasks
----------------------------------
listTestClusters - Lists all OpenSearch clusters configured for this project

Publishing tasks
----------------
generateMetadataFileForNebulaPublication - Generates the Gradle metadata file for publication 'nebula'.
generateMetadataFileForPluginZipPublication - Generates the Gradle metadata file for publication 'pluginZip'.
generatePomFileForNebulaPublication - Generates the Maven POM file for publication 'nebula'.
generatePomFileForPluginZipPublication - Generates the Maven POM file for publication 'pluginZip'.
publish - Publishes all publications produced by this project.
publishAllPublicationsToSnapshotsRepository - Publishes all Maven publications produced by this project to the Snapshots repository.
publishAllPublicationsToStagingRepository - Publishes all Maven publications produced by this project to the staging repository.
publishAllPublicationsToZipStagingRepository - Publishes all Maven publications produced by this project to the zipStaging repository.
publishNebulaPublicationToMavenLocal - Publishes Maven publication 'nebula' to the local Maven repository.
publishNebulaPublicationToSnapshotsRepository - Publishes Maven publication 'nebula' to Maven repository 'Snapshots'.
publishNebulaPublicationToStagingRepository - Publishes Maven publication 'nebula' to Maven repository 'staging'.
publishNebulaPublicationToZipStagingRepository - Publishes Maven publication 'nebula' to Maven repository 'zipStaging'.
publishPluginZipPublicationToMavenLocal - Publishes Maven publication 'pluginZip' to the local Maven repository.
publishPluginZipPublicationToSnapshotsRepository - Publishes Maven publication 'pluginZip' to Maven repository 'Snapshots'.
publishPluginZipPublicationToStagingRepository - Publishes Maven publication 'pluginZip' to Maven repository 'staging'.
publishPluginZipPublicationToZipStagingRepository - Publishes Maven publication 'pluginZip' to Maven repository 'zipStaging'.
publishToMavenLocal - Publishes all Maven publications produced by this project to the local Maven cache.

Verification tasks
------------------
check - Runs all checks.
diffCoverage - Builds coverage report only for modified code
jacocoTestCoverageVerification - Verifies code coverage metrics based on specified rules for the test task.
jacocoTestReport - Generates code coverage report for the test task.
precommit - Runs all non-test checks
spotlessApply - Applies code formatting steps to sourcecode in-place.
spotlessCheck - Checks that sourcecode satisfies formatting steps.
spotlessDiagnose
spotlessJava
spotlessJavaApply
spotlessJavaCheck
spotlessJavaDiagnose
spotlessLicense
spotlessLicenseApply
spotlessLicenseCheck
spotlessLicenseDiagnose
spotlessMisc
spotlessMiscApply
spotlessMiscCheck
spotlessMiscDiagnose
test - Runs the test suite.
yamlRestTest - Runs the REST tests against an external cluster

Rules
-----
Pattern: clean<TaskName>: Cleans the output files of a task.
Pattern: build<ConfigurationName>: Assembles the artifacts of a configuration.

To see all tasks and more detail, run gradlew tasks --all

To see more detail about a task, run gradlew help --task <task>

BUILD SUCCESSFUL in 49s

```

peterzhuamazon commented:
Approved @martin-gaievski  but note both UBI and SR will not have the necessary credentials to publish to snapshots maven yet, but this will unblock from non-snapshot mavenlocal which is used for Jenkins build.

Thanks.
